src/LuaUtils.cpp

```diff
@@ -176,122 +176,261 @@ static bool _import_core(lua_State *L, const std::string &importName)
+	std::string caller;
+	lua_Debug ar;
 
-	// get the cache of previously imported things
-	lua_getfield(L, LUA_REGISTRYINDEX, "Imports");
-	assert(lua_istable(L, -1));
+	// if have stack and caller in the stack
+	if (lua_getstack(L, 1, &ar) && lua_getinfo(L, "S", &ar) && ar.source)
+	{
+		int start = 0;
+		int end = strlen(ar.source);
 
-	// first check the cache for the explicit name (its there if it was
-	// previously imported from core)
-	std::string cacheName = importName;
-	lua_getfield(L, -1, cacheName.c_str());
-	if (!lua_isnil(L, -1)) {
-		// found it, use it
-		DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-		lua_remove(L, -2);
-		return 1;
+		if (ar.source[start] != '@' && end > 5)
+			start = 4;
+
+		if (ar.source[start] == '@') {
+			start++;
+
+			static char source[1024];
+			strncpy(source, &ar.source[start], end - start + 1);
+
+			for (end = end - start - 1; end >= 0; end--)
+			{
+				if (source[end] == '/') {
+					source[end] = '\0';
+					break;
+				}
+			}
+
+			caller = std::string(source);
+		}
 	}
-	lua_pop(L, 1);
-	DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
 
-	// next we're going to look in libs/
-	cacheName = FileSystem::JoinPath("libs", importName+".lua");
-	lua_getfield(L, -1, cacheName.c_str());
-	if (!lua_isnil(L, -1)) {
-		// found it, use it
-		DEBUG_PRINTF("import: found cached %s\n", cacheName.c_str());
-		lua_remove(L, -2);
-		return 1;
+	LUA_DEBUG_END(L, 0);
+	return caller;
+}
+
+/**
+* Finds field in table.
+* Require a existing table in the stack at index -1.
+* Returns true if field exists and not nil, puts value to lua stack.
+*/
+static bool get_cached(lua_State *L, const std::string& name)
+{
+	assert(L);
+	LUA_DEBUG_START(L);
+
+	bool found = false;
+
+	// try to get it
+	lua_getfield(L, -1, name.c_str());
+
+	// if got cached field
+	if (!lua_isnil(L, -1))
+	{
+		found = true;
+	}
+	else
+	{
+		// remove found "nil" from stack
+		lua_pop(L, 1);
 	}
-	lua_pop(L, 1);
-	DEBUG_PRINTF("import: not cached %s\n", cacheName.c_str());
 
-	RefCountedPtr<FileSystem::FileData> code = FileSystem::gameDataFiles.ReadFile(cacheName);
+	LUA_DEBUG_END(L, 0);
+	return found;
+}
 
-	// if its not found then we try to find the lib relative to the calling file
-	if (!code) {
-		DEBUG_PRINTF("import: not found %s, trying relative\n", cacheName.c_str());
+/**
+* Tries to load file.
+* Returns true if file exists and puts return of dofile to stack.
+*/
+static bool load_file(lua_State *L, const std::string& name)
+{
+	assert(L);
+	LUA_DEBUG_START(L);
 
-		lua_Debug ar;
-		lua_getstack(L, 1, &ar);
-		lua_getinfo(L, "S", &ar);
+	bool loaded = false;
+	RefCountedPtr<FileSystem::FileData> fileData = FileSystem::gameDataFiles.ReadFile(name);
 
-		if (ar.source) {
-			int start = 0;
-			int end = strlen(ar.source);
+	// if file exists
+	if (fileData)
+	{
+		pi_lua_dofile(L, *fileData, 1);
+		loaded = true;
+	}
 
-			if (ar.source[start] != '@' && end > 5)
-				start = 4;
+	LUA_DEBUG_END(L, 0);
```

Comment:
Two questions:
1/ Why the need for the loaded flag when you can just return a test on fileData ?
2/ Reading the function description, this stack check should be LUA_DEBUG_END(L, fileData?1:0)

Reply:
1. It also doing dofile(). Just simple structure of code to read. I can to edit this if one bool costs to much =)
2. Dont know enough LUA_DEBUG, im just copied that from existing code. In the next commit I will try to edit that.